packages/connect/src/core/index.ts

```diff
@@ -482,6 +488,37 @@ const onCall = async (message: IFrameCallMessage) => {
             createUiMessage(UI.REQUEST_PASSPHRASE_ON_DEVICE, { device: device.toMessageObject() }),
         );
     });
+    device.on(DEVICE.SAVE_STATE, (state: string) => {
+        storage.saveForOrigin(store => {
+            // Persist internal state only in case of core in popup
+            // Currently also only for webextension until we asses security implications
+            if (useCoreInPopup && env === 'webextension') {
+                return {
+                    ...store,
+                    preferredDevice: store.preferredDevice
+                        ? {
+                              ...store.preferredDevice,
+                              internalState: state,
+                              internalStateExpiration: new Date(Date.now() + 1000 * 60 * 15), // 15 minutes
```

Comment:
This is using Date object, I would rather use milliseconds like  below, since that is easier serializable and less potential issues and easier comparison with other dates.
```suggestion
                              internalStateExpiration: new Date(Date.now() + 1000 * 60 * 15).getTime(), // 15 minutes
```